From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tom Levy <tomlevy93@gmail.com>
Date: Wed, 18 Dec 2019 01:01:39 +0000
Subject: cherry-pick: Allow multiple team categories for self-registration

Replace the configuration option 'registration_category_name' with a
boolean property TeamCategory::$allow_self_registration. When it is
set to true on multiple categories, they are offered as choices during
registration.

(cherry picked from commit 696d33f8b37348e4b5e6b7169b476d8d77ad40d2)
---
 doc/manual/config-advanced.rst                | 12 ++--
 etc/db-config.yaml                            |  5 --
 .../Jury/TeamCategoryController.php           |  6 +-
 webapp/src/Controller/SecurityController.php  | 21 +++---
 webapp/src/Entity/TeamCategory.php            | 34 +++++++++
 webapp/src/Form/Type/TeamCategoryType.php     |  8 +++
 webapp/src/Form/Type/UserRegistrationType.php | 26 +++++++
 .../src/Migrations/Version20200131064449.php  | 72 +++++++++++++++++++
 webapp/src/Service/CheckConfigService.php     | 31 ++++++++
 webapp/templates/jury/team_category.html.twig |  4 ++
 10 files changed, 199 insertions(+), 20 deletions(-)
 create mode 100644 webapp/src/Migrations/Version20200131064449.php

diff --git a/doc/manual/config-advanced.rst b/doc/manual/config-advanced.rst
index d3f69fd4e..f9f8b33b1 100644
--- a/doc/manual/config-advanced.rst
+++ b/doc/manual/config-advanced.rst
@@ -113,11 +113,13 @@ will present the standard username/password login screen as a fallback.
 
 Self-registration
 `````````````````
-There is also a configuration option to allow teams to self-register with
-the system: ``registration_category_name``. When left empty, no self-registration
-is allowed; when filled with a category name, newly registered teams will
-be placed in this category. During registration, a team can specify their
-affiliation.
+Teams can be allowed to self-register with the system. Each team category can
+be set to allow registration or not. When none of the categories are set to
+allow, self-registration is disabled. When one category is set to allow,
+self-registration is enabled and newly registered teams will be placed in this
+category. When multiple categories are set to allow, teams can choose one of
+them during registration. Teams can also specify their affiliation during
+registration.
 
 Executables
 -----------
diff --git a/etc/db-config.yaml b/etc/db-config.yaml
index c28180296..be57cd7db 100644
--- a/etc/db-config.yaml
+++ b/etc/db-config.yaml
@@ -231,11 +231,6 @@
             default_value: ""
             public: true
             description: If not empty, enable teams and jury to send source code to this command. See admin manual for allowed arguments.
-        -   name: registration_category_name
-            type: string
-            default_value: ""
-            public: true
-            description: Team category for users that register themselves with the system. Self-registration is disabled if this field is left empty.
         -   name: data_source
             type: int
             default_value: 0
diff --git a/webapp/src/Controller/Jury/TeamCategoryController.php b/webapp/src/Controller/Jury/TeamCategoryController.php
index 14c4d5525..637f4e144 100644
--- a/webapp/src/Controller/Jury/TeamCategoryController.php
+++ b/webapp/src/Controller/Jury/TeamCategoryController.php
@@ -83,6 +83,7 @@ class TeamCategoryController extends BaseController
             'name' => ['title' => 'name', 'sort' => true],
             'num_teams' => ['title' => '# teams', 'sort' => true],
             'visible' => ['title' => 'visible', 'sort' => true],
+            'allow_self_registration' => ['title' => 'self-registration', 'sort' => true],
         ];
 
         // Insert external ID field when configured to use it
@@ -124,8 +125,9 @@ class TeamCategoryController extends BaseController
                 ];
             }
 
-            $categorydata['num_teams'] = ['value' => $teamCategoryData['num_teams']];
-            $categorydata['visible']   = ['value' => $teamCategory->getVisible() ? 'yes' : 'no'];
+            $categorydata['num_teams']               = ['value' => $teamCategoryData['num_teams']];
+            $categorydata['visible']                 = ['value' => $teamCategory->getVisible() ? 'yes' : 'no'];
+            $categorydata['allow_self_registration'] = ['value' => $teamCategory->getAllowSelfRegistration() ? 'yes' : 'no'];
 
             $team_categories_table[] = [
                 'data' => $categorydata,
diff --git a/webapp/src/Controller/SecurityController.php b/webapp/src/Controller/SecurityController.php
index fbfb8a2a4..29d363845 100644
--- a/webapp/src/Controller/SecurityController.php
+++ b/webapp/src/Controller/SecurityController.php
@@ -84,13 +84,12 @@ class SecurityController extends AbstractController
         $response = new Response();
         $response->headers->set('X-Login-Page', $this->generateUrl('login'));
 
-        $registrationCategoryName = $this->config->get('registration_category_name');
-        $registrationCategory     = $em->getRepository(TeamCategory::class)->findOneBy(['name' => $registrationCategoryName]);
+        $selfRegistrationCategoriesCount = $em->getRepository(TeamCategory::class)->count(['allow_self_registration' => 1]);
 
         return $this->render('security/login.html.twig', array(
             'last_username' => $lastUsername,
             'error' => $error,
-            'allow_registration' => $registrationCategory !== null,
+            'allow_registration' => $selfRegistrationCategoriesCount !== 0,
             'allowed_authmethods' => $authmethods,
             'auth_xheaders_present' => $request->headers->get('X-DOMjudge-Login'),
             'auth_ipaddress_users' => $auth_ipaddress_users,
@@ -116,11 +115,10 @@ class SecurityController extends AbstractController
             return $this->redirect($this->generateUrl('root'));
         }
 
-        $em                       = $this->getDoctrine()->getManager();
-        $registrationCategoryName = $this->config->get('registration_category_name');
-        $registrationCategory     = $em->getRepository(TeamCategory::class)->findOneBy(['name' => $registrationCategoryName]);
+        $em                              = $this->getDoctrine()->getManager();
+        $selfRegistrationCategoriesCount = $em->getRepository(TeamCategory::class)->count(['allow_self_registration' => 1]);
 
-        if ($registrationCategory === null) {
+        if ($selfRegistrationCategoriesCount === 0) {
             throw new HttpException(400, "Registration not enabled");
         }
 
@@ -138,13 +136,20 @@ class SecurityController extends AbstractController
 
             $teamName = $registration_form->get('teamName')->getData();
 
+            if ($selfRegistrationCategoriesCount === 1) {
+                $teamCategory = $em->getRepository(TeamCategory::class)->findOneBy(['allow_self_registration' => 1]);
+            } else {
+                // $selfRegistrationCategoriesCount > 1, 'teamCategory' field exists
+                $teamCategory = $registration_form->get('teamCategory')->getData();
+            }
+
             // Create a team to go with the user, then set some team attributes
             $team = new Team();
             $user->setTeam($team);
             $team
                 ->addUser($user)
                 ->setName($teamName)
-                ->setCategory($registrationCategory)
+                ->setCategory($teamCategory)
                 ->setComments('Registered by ' . $this->dj->getClientIp() . ' on ' . date('r'));
 
             if ($this->dj->dbconfig_get('show_affiliations', true)) {
diff --git a/webapp/src/Entity/TeamCategory.php b/webapp/src/Entity/TeamCategory.php
index 8fc45b2e9..3c0976f76 100644
--- a/webapp/src/Entity/TeamCategory.php
+++ b/webapp/src/Entity/TeamCategory.php
@@ -77,6 +77,16 @@ class TeamCategory extends BaseApiEntity
      */
     private $visible = true;
 
+    /**
+     * @var boolean
+     * @ORM\Column(type="boolean", name="allow_self_registration",
+     *     options={"comment"="Are self-registered teams allowed to choose this category?",
+     *              "default"="0"},
+     *     nullable=false)
+     * @Serializer\Exclude()
+     */
+    private $allow_self_registration = false;
+
     /**
      * @ORM\OneToMany(targetEntity="Team", mappedBy="category")
      * @Serializer\Exclude()
@@ -225,6 +235,30 @@ class TeamCategory extends BaseApiEntity
         return $this->visible;
     }
 
+    /**
+     * Set allowSelfRegistration
+     *
+     * @param boolean $allowSelfRegistration
+     *
+     * @return TeamCategory
+     */
+    public function setAllowSelfRegistration($allowSelfRegistration)
+    {
+        $this->allow_self_registration = $allowSelfRegistration;
+
+        return $this;
+    }
+
+    /**
+     * Get allowSelfRegistration
+     *
+     * @return boolean
+     */
+    public function getAllowSelfRegistration()
+    {
+        return $this->allow_self_registration;
+    }
+
     /**
      * Add team
      *
diff --git a/webapp/src/Form/Type/TeamCategoryType.php b/webapp/src/Form/Type/TeamCategoryType.php
index 704c0c22d..4f3749c31 100644
--- a/webapp/src/Form/Type/TeamCategoryType.php
+++ b/webapp/src/Form/Type/TeamCategoryType.php
@@ -32,6 +32,14 @@ class TeamCategoryType extends AbstractType
                 'No' => false,
             ],
         ]);
+        $builder->add('allow_self_registration', ChoiceType::class, [
+            'label' => 'Allow self-registration',
+            'expanded' => true,
+            'choices' => [
+                'Yes' => true,
+                'No' => false,
+            ],
+        ]);
         $builder->add('save', SubmitType::class);
     }
 
diff --git a/webapp/src/Form/Type/UserRegistrationType.php b/webapp/src/Form/Type/UserRegistrationType.php
index d32d2d821..c335e8d32 100644
--- a/webapp/src/Form/Type/UserRegistrationType.php
+++ b/webapp/src/Form/Type/UserRegistrationType.php
@@ -4,10 +4,12 @@ namespace App\Form\Type;
 
 use App\Entity\Team;
 use App\Entity\TeamAffiliation;
+use App\Entity\TeamCategory;
 use App\Entity\User;
 use App\Service\DOMJudgeService;
 use App\Utils\Utils;
 use Doctrine\ORM\EntityManagerInterface;
+use Doctrine\ORM\EntityRepository;
 use Symfony\Bridge\Doctrine\Form\Type\EntityType;
 use Symfony\Component\Form\AbstractType;
 use Symfony\Component\Form\Extension\Core\Type\ChoiceType;
@@ -92,6 +94,30 @@ class UserRegistrationType extends AbstractType
                 'mapped' => false,
             ]);
 
+        $selfRegistrationCategoriesCount = $this->em->getRepository(TeamCategory::class)->count(['allow_self_registration' => 1]);
+        if ($selfRegistrationCategoriesCount > 1) {
+            $builder
+                ->add('teamCategory', EntityType::class, [
+                    'class' => TeamCategory::class,
+                    'label' => false,
+                    'mapped' => false,
+                    'choice_label' => 'name',
+                    'placeholder' => '-- Select category --',
+                    'query_builder' => function (EntityRepository $er) {
+                        return $er
+                            ->createQueryBuilder('c')
+                            ->where('c.allow_self_registration = 1')
+                            ->orderBy('c.sortorder');
+                    },
+                    'attr' => [
+                        'placeholder' => 'Category',
+                    ],
+                    'constraints' => [
+                        new NotBlank(),
+                    ],
+                ]);
+        }
+
         if ($this->config->get('show_affiliations')) {
             $countries = [];
             foreach (Utils::ALPHA3_COUNTRIES as $alpha3 => $country) {
diff --git a/webapp/src/Migrations/Version20200131064449.php b/webapp/src/Migrations/Version20200131064449.php
new file mode 100644
index 000000000..e0731b1ad
--- /dev/null
+++ b/webapp/src/Migrations/Version20200131064449.php
@@ -0,0 +1,72 @@
+<?php
+
+declare(strict_types=1);
+
+namespace DoctrineMigrations;
+
+use App\Entity\Configuration;
+use App\Entity\TeamCategory;
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
+use Symfony\Component\DependencyInjection\ContainerAwareInterface;
+use Symfony\Component\DependencyInjection\ContainerAwareTrait;
+
+final class Version20200131064449 extends AbstractMigration implements ContainerAwareInterface
+{
+    use ContainerAwareTrait;
+
+    public function getDescription() : string
+    {
+        return 'replace configuration option registration_category_name with a boolean field for each category';
+    }
+
+    public function up(Schema $schema) : void
+    {
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() !== 'mysql', 'Migration can only be executed safely on \'mysql\'.');
+
+        // Note: can't use ConfigurationService::get on 'registration_category_name' because the specification has been removed from db-config.yaml
+        $em = $this->container->get('doctrine')->getManager();
+        $registrationCategoryNameConfig = $em->getRepository(Configuration::class)->findOneBy(['name' => 'registration_category_name']);
+        if ($registrationCategoryNameConfig) {
+            $registrationCategoryName = $registrationCategoryNameConfig->getValue();
+        } else {
+            $registrationCategoryName = '';
+        }
+
+        $this->addSql('ALTER TABLE team_category ADD allow_self_registration TINYINT(1) DEFAULT \'0\' NOT NULL COMMENT \'Are self-registered teams allowed to choose this category?\'');
+
+        if ($registrationCategoryName !== '') {
+            $this->addSql(
+                'UPDATE team_category SET allow_self_registration = 1 WHERE name = :name',
+                ['name' => $registrationCategoryName]
+            );
+        }
+        $this->addSql("DELETE FROM configuration WHERE name = 'registration_category_name'");
+    }
+
+    public function down(Schema $schema) : void
+    {
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() !== 'mysql', 'Migration can only be executed safely on \'mysql\'.');
+
+        $em = $this->container->get('doctrine')->getManager();
+        $selfRegistrationCategories = $em->getRepository(TeamCategory::class)->findBy(
+            ['allow_self_registration' => 1],
+            ['sortorder' => 'ASC']
+        );
+
+        $this->warnIf(
+            count($selfRegistrationCategories) > 1,
+            sprintf('Team categories for self-registered teams were %s. Only first will be kept.',
+                implode(', ', array_map(function($category) {
+                    return $category->getName();
+                }, $selfRegistrationCategories)))
+        );
+
+        $this->addSql(
+            "INSERT INTO configuration (name, value) VALUES ('registration_category_name', :value)",
+            ['value' => empty($selfRegistrationCategories) ? '""' : json_encode($selfRegistrationCategories[0]->getName())]
+        );
+
+        $this->addSql('ALTER TABLE team_category DROP allow_self_registration');
+    }
+}
diff --git a/webapp/src/Service/CheckConfigService.php b/webapp/src/Service/CheckConfigService.php
index 9931632f4..9e8b22871 100644
--- a/webapp/src/Service/CheckConfigService.php
+++ b/webapp/src/Service/CheckConfigService.php
@@ -9,6 +9,7 @@ use App\Entity\Language;
 use App\Entity\Problem;
 use App\Entity\Team;
 use App\Entity\TeamAffiliation;
+use App\Entity\TeamCategory;
 use App\Entity\Testcase;
 use App\Entity\User;
 use App\Utils\Utils;
@@ -111,6 +112,7 @@ class CheckConfigService
         $teams = [
             'affiliations' => $this->checkAffiliations(),
             'teamdupenames' => $this->checkTeamDuplicateNames(),
+            'selfregistration' => $this->checkSelfRegistration(),
         ];
 
         $results['Teams'] = $teams;
@@ -657,6 +659,35 @@ class CheckConfigService
             'desc' => $desc];
     }
 
+    public function checkSelfRegistration()
+    {
+        $result = 'O';
+        $desc = '';
+
+        $selfRegistrationCategories = $this->em->getRepository(TeamCategory::class)->findBy(
+            ['allow_self_registration' => 1],
+            ['sortorder' => 'ASC']
+        );
+        if (count($selfRegistrationCategories) === 0) {
+            $desc .= "Self-registration is disabled.\n";
+        } else {
+            $desc .= "Self-registration is enabled.\n";
+            if (count($selfRegistrationCategories) === 1) {
+                $desc .= sprintf("Team category for self-registered teams: %s.\n",
+                    $selfRegistrationCategories[0]->getName());
+            } else {
+                $desc .= sprintf("Team categories allowed for self-registered teams: %s.\n",
+                    implode(', ', array_map(function($category) {
+                        return $category->getName();
+                    }, $selfRegistrationCategories)));
+            }
+        }
+
+        return ['caption' => 'Self-registration',
+            'result' => $result,
+            'desc' => $desc];
+    }
+
     public function checkAllExternalIdentifiers()
     {
         // Get all entity classes
diff --git a/webapp/templates/jury/team_category.html.twig b/webapp/templates/jury/team_category.html.twig
index 631c2c428..b94e82ea8 100644
--- a/webapp/templates/jury/team_category.html.twig
+++ b/webapp/templates/jury/team_category.html.twig
@@ -37,6 +37,10 @@
                     <th>Visible</th>
                     <td>{{ teamCategory.visible | printYesNo }}</td>
                 </tr>
+                <tr>
+                    <th>Allow self-registration</th>
+                    <td>{{ teamCategory.allowSelfRegistration | printYesNo }}</td>
+                </tr>
             </table>
         </div>
     </div>
-- 
2.20.1

